### PR TITLE
Select issue under cursor and bulk update

### PR DESCRIPTION
**Pull request by MarkusAmshove**
Hi,

thanks for your awesome plugin!

I've added a functionality which lets you do a "bulk update" of an issue, where the issue id is selected via cursor.

This way I can use it from within my TODO list.

Lets say I have the following issue:

image

I can now update the issue (status, assignee, write a comment) within one step:

pr

I couldn't really decide on how to name it, if you'd like to have a different name I'll do it :-)

### CODE REVIEW

**Comment by MarkusAmshove** [comment]
Having the current commands in mind, I think it might make more sense to split the command up.

New commands:
* Select issue under cursor
* Bulk update issue

That way the ` select issue under cursor`  could be used instead of entering an issue id for ` List actions for issue` and the bulk update could just be another action in `actions for issue`

File: src/extension.ts [hunk 1]
```diff
@@ -137,6 +138,67 @@ export function activate(context: vscode.ExtensionContext) {
...
+        );
+        if(!assigneeChoice) {
+            return;
+        }
+
+        const desiredAssignee = assigneeChoice.assignee;
+        const message = await vscode.window.showInputBox({ placeHolder: "Message" });
+
+        const bulkUpdate = new BulkUpdate(issueId, message, desiredAssignee, desiredStatus);
+        const updateResult = await redmine.applyBulkUpdate(bulkUpdate);
```
Comment (rozpuszczalny) [inline src/extension.ts]: Same as line 153, no error checking from redmine server.

**Comment by MarkusAmshove** [comment]
Thank you for the review, I'll split the functionality as suggested and keep your comments in mind.

I'm not that experienced with JS/TS stuff, that's why I didn't know how to handle the error cases from a Promise correctly.
That's why I implemented the the validation (reread issue and confirm updates) in the first place  :-)

Should I split those two into separate PRs or are two features in one PR okay for you?

**Comment by MarkusAmshove** [comment]
It does partial update, e.g updating the assignee but not status.
So I thought rereading and notifying the user about a partial update would be good UX.

I would like to keep the reread and change the user facing warning message to "Issue partially updated" with the info that e.g. the status couldn't be changed

**Comment by MarkusAmshove** [comment]
The functionalities are now split up.

* There is no more `do X for issue under cursor` 
* New command `List actions for issue under cursor`
* Added `QuickUpdate` to list of actions

Screenshot of the partial validation, where the workflow in Redmine forbid changing from `In Progress` to `Backlog`:

partial

Now I'm ready for comments on error handling :-)

File: src/controllers/issue-controller.ts [hunk 2]
```diff
@@ -82,6 +83,58 @@ export class IssueController {
...
+
+        const desiredAssignee = assigneeChoice.assignee;
+        const message = await vscode.window.showInputBox({ placeHolder: "Message" });
+
+        const bulkUpdate = new QuickUpdate(this.issue.id, message, desiredAssignee, desiredStatus);
+        const updateResult = await this.redmine.applyQuickUpdate(bulkUpdate);
+        if(updateResult.isSuccessful()) {
+            vscode.window.showInformationMessage("Issue updated");
+        } else {
+            vscode.window.showErrorMessage("Issue updated partially", ...updateResult.differences);
```
Comment (rozpuszczalny) [inline src/controllers/issue-controller.ts]: Diffrences show up as buttons in error message. It should be done like this:
```typescript
vscode.window.showErrorMessage(
  `Issue updated partially; problems: \n ${updateResult.differences.join('\n')}`);
```

File: src/controllers/issue-controller.ts [hunk 3]
```diff
@@ -82,6 +83,58 @@ export class IssueController {
...
+        );
+        if(!assigneeChoice) {
+            return;
+        }
+
+        const desiredAssignee = assigneeChoice.assignee;
+        const message = await vscode.window.showInputBox({ placeHolder: "Message" });
+
+        const bulkUpdate = new QuickUpdate(this.issue.id, message, desiredAssignee, desiredStatus);
+        const updateResult = await this.redmine.applyQuickUpdate(bulkUpdate);
```
Comment (rozpuszczalny) [inline src/controllers/issue-controller.ts]: There should be an error handler. If `doRequest` in `applyQuickUpdate` throws an error (e.g. not authorized, internal server error, no connection etc.), it won't be catched by re-read mechanism in `applyQuickUpdate`.

File: src/controllers/issue-controller.ts [hunk 4]
```diff
@@ -82,6 +83,58 @@ export class IssueController {
         });
     }
 
+    private async quickUpdate() {
+        const memberships = await this.redmine.getMemberships(this.issue.project.id);
+        const possibleStatuses = await this.redmine.getIssueStatusesTyped();
```
Comment (rozpuszczalny) [inline src/controllers/issue-controller.ts]: Above two lines should have error catching too.

**Comment by MarkusAmshove** [comment]
Should be good to go now :-)

**Comment by rozpuszczalny** [comment]
LGTM 👍

**Pull request merged**